[상품] DB대신 MAP을 이용한 API 구현

minnnkyeng opened the pull request:
## 구현 내용
ResponseEntity를 이용한 상품 도메인 기능 구현


## 주요 구현 내용
- 기존에 구현한 기능에 ResponseEntity를 이용 및 각 기능에 대한 예외 구현 (ProrductController)
[src/main/java/com/basecamp/ticketpick/product/ProductController.java](https://github.com/minnnkyeng/ticketPick/blob/dev-mapDB/src/main/java/com/basecamp/ticketpick/product/ProductController.java)

- 엔티티와 구분하기 위한 DTO 구현 (ProductService.updateProduct)
[src/main/java/com/basecamp/ticketpick/product/ProductService.java](https://github.com/minnnkyeng/ticketPick/blob/dev-mapDB/src/main/java/com/basecamp/ticketpick/product/ProductService.java)


## 이슈
- 에러 : 상품 수정 구현 시,  수정 요청하지 않은 필드에 대해 NULL값으로 처리됨
- 해결방안
: 상품 엔티티와 DTO로 구분.
: 요청 데이터는 DTO에 저장하고 수정하려는 상품의 기존 데이터와 요청 데이터를 비교하여 요청 데이터를 기존 데이터에 세팅하였음.
- 고민사항 : DTO 사용이 적절한가. 각 기능별 HTTP상태코드 사용이 적절한가.

### 다음 PR 목표
1. 상품 엔티티와 DTO를 구분하면서 엔티티에 setter를 구현하게 되었는데 권장사항대로 엔티티에 setter를 없애고 다른 방안으로 구현
2. JPA 구현
3. 상품과 관련된 엔티티 구현 및 기능 추가

minnnkyeng committed:
fix : ResponseEntity를 이용한 기능 구현

- 기존 기능에 ResponseEntity를 이용
- 도메인과 구분하기 위한 DTO 구현
- 각 기능별 예외에 따른 코드 추가

* 기능구현
getAllProduct() : 상품 전체조회
getProduct(int) : 상품 상세조회
saveProduct(Product) : 상품 등록
updateProduct(int, ProductRequestDTO) : 상품 수정
deleteProduct(int) : 상품 삭제

* files
ProductController.java
ProductService.java
ProductRepository.java
ProductMapRepository.java
ProductRequestDTO.java
Product.java

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductService.java:
```diff
@@ -10,23 +12,38 @@ public ProductService(ProductRepository productRepository) {
     }
 
-    public String saveProduct(Product product) {
+    public int saveProduct(Product product) {
```
get - set
find - save
이름 세트 생각해봐주시면 더 좋을 듯!

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/Product.java:
말씀하신 것처럼 setter보단 builder, Mapper를 활용해보시는 걸 추천드립니다!

ssongCoding replied:
단, 수정용으로는 Entity에 setter가 필요할 수 있어서 이건 추후 고민해보실게요!

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductController.java:
```diff
@@ -11,27 +15,55 @@ public ProductController(ProductService productService) {
     }
 
     @GetMapping("/products")
```
공통된 URL은 클래스 단에서 묶어주시죠!

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductController.java:
```diff
@@ -11,27 +15,55 @@ public ProductController(ProductService productService) {
+            return ResponseEntity.badRequest().body("Request data is null.");
+        }
+        int id = productService.saveProduct(product);
+        return ResponseEntity.created(URI.create("/products/"+id)).build();
```
create로 분할해주신 것 쿠우우울👍🏻

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductController.java:
```diff
@@ -11,27 +15,55 @@ public ProductController(ProductService productService) {
-        return productService.getProduct(id);
+    public ResponseEntity<?> getProduct(@PathVariable int id) {
+        Product product = productService.getProduct(id);
+        if(product==null){
+            return ResponseEntity.ok("There is not product, id="+id);
+        }
```
product 객체가 null이라면, 컨트롤러가 알기 전에 서비스가 먼저 알았겠죠? 서비스 단에서도 처리 해주셔야 할듯 합니다!

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductService.java:
```diff
@@ -10,23 +12,38 @@ public ProductService(ProductRepository productRepository) {
-    public String updateProduct(Product product) {
-        return productRepository.updateProduct(product);
+    public void updateProduct(Product product, ProductRequestDTO productRequestDTO) {
+        if(productRequestDTO.getName()!=null){
+            product.setName(productRequestDTO.getName());
+        }
+        if(productRequestDTO.getDescription()!=null){
+            product.setDescription(productRequestDTO.getDescription());
+        }
+        if(productRequestDTO.getPrice()<0){
+            product.setPrice(productRequestDTO.getPrice());
+        }
+        if(productRequestDTO.getDate()!=null){
+            product.setDate(productRequestDTO.getDate());
+        }
+        if(productRequestDTO.getVenue()!=null){
+            product.setVenue(productRequestDTO.getVenue());
+        }
+        productRepository.updateProduct(product);
```
이히히히 if가 너무 많아요! 고도화해보시죠!

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductMapRepository.java:
```diff
@@ -0,0 +1,32 @@
+    private Map<Integer, Product> db = new HashMap<>();
+    private int id = 0;
+
+    public Map<Integer, Product> getAllProduct() { return db; }
```
ㅋㅋㅋㅋ디비를 통으로 날려버리는 그대란 사람..🔥 화끈해버렸다

서비스나 컨트롤러에서 내부 데이터 수정이 필요하다면, Map을 통째로 던지면 서비스든 누구든 map -> stream -> list로 바꿔주는 것이 좋습니다. 또한 클라이언트도 리스트로 받는게 좋아서! 리스트화 해보실까요?

ssongCoding commented on src/main/java/com/basecamp/ticketpick/product/ProductController.java:
```diff
@@ -11,27 +15,55 @@ public ProductController(ProductService productService) {
-        return productService.deleteProduct(id);
+    public ResponseEntity<?> deleteProduct(@PathVariable int id) {
+        Product product = productService.getProduct(id);
+        if(product==null){
+            return ResponseEntity.ok("There is not product, id="+id);
+        }
+        if(id < 1){
+            return ResponseEntity.badRequest().body("Product id is under 1");
+        }
```
Exception을 안쓰고도 우린 예외 처리를 고민했다는 흔적이 여기 잔뜩 남아있네요.

클라이언트한테는 민경님께서 구현하신 것처럼 ok, badrequest 등 알려주는 것이 좋습니다👍🏻
백엔드 내부적으로는 Exception을 추가해주시면 더 좋겠네요!

minnnkyeng committed:
fix : 공통된 URL 묶기

- 공통된 URL을 클래스 레벨에서 작성(@RequestMapping)

* files
ProductController.java

minnnkyeng committed:
fix : map DB 리스트화, 메소드명 변경

- findAllProduct()의 반환타입을 map -> list로 변경
- 메소드명 변경
(getAllProduct -> findAllProduct, getProduct -> findProduct)

* files
ProductController.java
ProductService.java
ProductRepository.java
ProductMapRepository.java
